warehouse/query-core/src/main/java/datawave/query/postprocessing/tf/DocumentKeysFunction.java

```diff
@@ -0,0 +1,291 @@
+import java.util.Set;
+import java.util.TreeSet;
+
+import static datawave.query.jexl.functions.ContentFunctions.CONTENT_ADJACENT_FUNCTION_NAME;
+import static datawave.query.jexl.functions.ContentFunctions.CONTENT_PHRASE_FUNCTION_NAME;
+import static datawave.query.jexl.functions.ContentFunctions.CONTENT_SCORED_PHRASE_FUNCTION_NAME;
+import static datawave.query.jexl.functions.ContentFunctions.CONTENT_WITHIN_FUNCTION_NAME;
+import static datawave.query.jexl.functions.ContentFunctions.TERM_OFFSET_MAP_JEXL_VARIABLE_NAME;
+
+/**
+ * A function that intersects document keys prior to fetching term offsets.
+ * <p>
+ * This only works on positive phrase queries -- that is, no guarantee is made for negated phrases
+ */
+public class DocumentKeysFunction {
+    
+    private static final Logger log = Logger.getLogger(DocumentKeysFunction.class);
+    
+    private final Multimap<String,ContentFunction> contentFunctions = LinkedListMultimap.create();
+    
+    /**
+     * @param config
+     *            a {@link TermFrequencyConfig}
+     */
+    public DocumentKeysFunction(TermFrequencyConfig config) {
+        populateContentFunctions(config.getScript());
+    }
+    
+    protected void populateContentFunctions(JexlNode node) {
+        
+        Multimap<String,Function> functions = TermOffsetPopulator.getContentFunctions(node);
+        for (String key : functions.keySet()) {
+            Collection<Function> coll = functions.get(key);
+            for (Function f : coll) {
+                ContentFunction contentFunction = transformFunction(f);
+                contentFunctions.put(contentFunction.getField(), contentFunction);
+            }
+        }
+    }
+    
+    private ContentFunction transformFunction(Function function) {
+        
+        boolean specialCase = false;
+        int index = findValueStartIndex(function);
+        
+        // Parse the values first. Might have to lookup fields by value.
+        TreeSet<String> values = getValuesFromArgs(function.args(), index);
+        
+        TreeSet<String> fields;
+        if (specialCase) {
+            // field(s) were not present in the function, and cannot lookup fields by value.
+            throw new IllegalStateException("cannot lookup fields by value");
+        } else {
+            // function's first arg is the field, or fields in the form (FIELD_A || FIELD_B)
+            fields = parseField(function.args().get(0));
+        }
+        
+        return new ContentFunction(fields.first(), values);
+    }
+    
+    // functions may take different forms..
+    // within = {field, number, termOffsetMap, terms...}
+    // within = {number, termOffsetMap, terms...}
+    // adjacent = {field, termOffsetMap, terms...}
+    // adjacent = {termOffsetMap, terms...}
+    // phrase = {field, termOffsetMap, terms...}
+    // phrase = {termOffsetMap, terms...}
+    // scoredPhrase = {-1.5, termOffsetMap,'boy','car'}
+    private int findValueStartIndex(Function function) {
```

Comment:
This seems like something the functions descriptor should be able to tell us.

Reply:
I'll take a cut at using the descriptors

Reply:
Switched to using the ContentFunctionsDescriptor